### PR TITLE
Introduce data/recorder-driven testing

### PR DESCRIPTION
**Pull request by irfansharif**
TestDatadriven makes use of datadriven and recorder to capture all
operations executed by individual dev invocations. The testcases are
defined under testdata/\*, where each test files corresponds to a
recording capture found in testdata/recording/\*.

Datadriven divvies up these files as subtests, so individual "files" are
runnable through:

    go test -run TestDatadriven/<fname>

To update the test file with new capture data, try:

    go test -run TestDatadriven/<fname> -rewrite -record \
            [-from-checkout=<path to crdb checkout>]

The -rewrite flag rewrites what is found under testdata/<fname>, while
the -record flag rewrites what is found under
testdata/recording/<fname>. Recordings are used to mock out "external
system" behavior. During these test runs (unless -record is specified),
attempts to shell out to bazel are intercepted and responses are
constructed using recorded data. The same is true for attempts to run os
operations (like creating/removing/symlinking filepaths).

---

For an example, consider testdata/generate:

    dev gen bazel
    ----
    bazel run //:gazelle -- update-repos -from_file=go.mod <...>
    getwd
    bazel run //pkg/cmd/generate-test-suites --run_under cd go/src/githu<...>
    writefile: go/src/github.com/cockroachdb/cockroach/pkg/BUILD.bazel
    bazel run @cockroach//:gazelle --color=yes

And the corresponding recording, testdata/recording/generate:

    bazel run //:gazelle -- update-repos -from_file=go.mod -build_file_<...>
    ----

    getwd
    ----
    go/src/github.com/cockroachdb/cockroach

    bazel run //pkg/cmd/generate-test-suites --run_under cd go/src/github<...>
    ----
    ----
    # Code generated by generate-test-suites, DO NOT EDIT.
    # gazelle:proto_strip_import_prefix /pkg

    ALL_TESTS = [
        "//pkg/util/tracing:tracing_test",
    ]

    test_suite(
        name = "small_tests",
        tags = [
            "-broken_in_bazel",
            "small",
        ],
        tests = ALL_TESTS,
    )
    ----
    ----

    writefile: go/src/github.com/cockroachdb/cockroach/pkg/BUILD.bazel
    ----

    bazel run @cockroach//:gazelle --color=yes
    ----

Once the recordings are captured, they can be edited and maintained by
hand. An example of where we might want to do this is for recordings for
commands that generate copious amounts of output (`bazel aquery` is an
instance of this; `generate-test-suites` above is another). It suffices
for us to trim the recording down by hand, and make sure we don't
re-record over it (by inspecting git diffs during review).

---

In summary: the traces for all operations attempted as part of test run
are captured within testdata/<fname> and the mocked out responses for
each "external" command can be found under testadata/recording/<fname>.
The former is updated by specifying -rewrite, the latter is updated by
specifying -record (and optionally -from-checkout).

### CODE REVIEW

**Review by rickystewart — APPROVED** [review]
Works for me, thanks for finding a workable approach to unit testing.

> Once the recordings are captured, they can be edited and maintained by
> hand. An example of where we might want to do this is for recordings for
> commands that generate copious amounts of output (bazel aquery is an
> instance of this; generate-test-suites above is another). It suffices
> for us to trim the recording down by hand, and make sure we don't
> re-record over it (by inspecting git diffs during review).

Seems fine for now -- I worry about the scalability of this, but `dev` isn't going to get THAT big, right? :)

**Comment by irfansharif** [comment]
TFTR!
